gremlin-groovy/src/main/java/org/apache/tinkerpop/gremlin/groovy/jsr223/ParameterizedGroovyTranslator.java

```diff
@@ -0,0 +1,340 @@
+import org.apache.tinkerpop.gremlin.util.function.Lambda;
+
+import java.util.*;
+import java.util.stream.Collectors;
+
+/**
+ * Converts bytecode to a Parameterized Groovy string of Gremlin.
+ *
+ * @author Marko A. Rodriguez (http://markorodriguez.com)
+ * @author Stephen Mallette (http://stephen.genoprime.com)
+ * @author Stark Arya (sandszhou.zj@alibaba-inc.com)
+ */
+
+public class ParameterizedGroovyTranslator implements ScriptTranslator {
+    private final String traversalSource;
+    private final TypeTranslator typeTranslator;
+
+    private ParameterizedGroovyTranslator(final String traversalSource, TypeTranslator typeTranslator) {
+        this.traversalSource = traversalSource;
+        this.typeTranslator = typeTranslator;
+    }
+
+    public static final ParameterizedGroovyTranslator of(final String traversalSource) {
+        return of(traversalSource,null);
+    }
+
+    public static final ParameterizedGroovyTranslator of(final String traversalSource, final TypeTranslator typeTranslator) {
+        return new ParameterizedGroovyTranslator(traversalSource,
+                Optional.ofNullable(typeTranslator).orElseGet(DefaultParameterizedTypeTranslator::new));
+    }
+
+    //////
+    public ParameterizedGroovyResult parameterizedTranslate(final Bytecode bytecode) {
+        String groovy = typeTranslator.apply(traversalSource, bytecode).toString();
+
+        Map<String, Object> parameters = (typeTranslator instanceof  DefaultParameterizedTypeTranslator) ?
+                ((DefaultParameterizedTypeTranslator) typeTranslator).getParameters() : new HashMap<>(0);
+
+        return new ParameterizedGroovyResult(groovy, parameters);
+    }
+
+    @Override
+    public String translate(final Bytecode bytecode) {
+        return typeTranslator.apply(traversalSource, bytecode).toString();
+    }
+
+    @Override
+    public String getTargetLanguage() {
+        return "gremlin-groovy";
+    }
+
+    @Override
+    public String toString() {
+        return StringFactory.translatorString(this);
+    }
+
+    @Override
+    public String getTraversalSource() {
+        return this.traversalSource;
+    }
+
+    /**
+     * Parameterized Script Result
+     */
+    public final class ParameterizedGroovyResult {
+        private final String  script;
+        private final Map<String,Object> parameters;
+
+        public ParameterizedGroovyResult(String script, Map<String,Object> parameters) {
+            this.script = script;
+            this.parameters = Collections.unmodifiableMap(parameters);
+        }
+
+        public String getScript() {
+            return script;
+        }
+
+        public Map<String,Object> getParameters() {
+            return parameters;
+        }
+    }
+
+    /**
+     * Performs standard type translation for the TinkerPop types to Parameterized Groovy.
+     */
+    public static class DefaultParameterizedTypeTranslator implements TypeTranslator {
```

Comment:
It would be nice if this could all be implemented as a simple `TypeTranslator` because then we wouldn't need this special  `Translator` implementation. Still thinking about how this might all "just work" so I could perhaps suggest more changes after these, but for now, I don't think that the constructor for `ParameterizedGroovyTranslator` should have a construction that take a simple `TypeTranslator`:

https://github.com/apache/tinkerpop/pull/1157/files#diff-1ca5e24d7542dff7f3f8fbbca51c91faR61

because for it to extract parameters it needs a special kind of `TypeTranslator` (i.e. one that captures parameters like this one). So, you might need a new `ParameterizedTypeTranslator` interface to accomplish that. I wish you didn't though. Overall this needs a bit more thought to get t right.